usaspending_api/search/v2/views/spending_by_award.py

```diff
@@ -0,0 +1,259 @@
+    award_idv_mapping,
+    loan_award_mapping,
+    non_loan_assistance_award_mapping,
+)
+from usaspending_api.common.api_versioning import api_transformations, API_TRANSFORM_FUNCTIONS
+from usaspending_api.common.cache_decorator import cache_response
+from usaspending_api.common.exceptions import InvalidParameterException, UnprocessableEntityException
+from usaspending_api.common.helpers.orm_helpers import obtain_view_from_award_group
+from usaspending_api.common.validator.award_filter import AWARD_FILTER
+from usaspending_api.common.validator.pagination import PAGINATION
+from usaspending_api.common.validator.tinyshield import TinyShield
+
+
+@api_transformations(api_version=settings.API_VERSION, function_list=API_TRANSFORM_FUNCTIONS)
+class SpendingByAwardVisualizationViewSet(APIView):
+    """
+    This route takes award filters and fields, and returns the fields of the filtered awards.
+    """
+
+    endpoint_doc = "usaspending_api/api_docs/api_documentation/advanced_award_search/spending_by_award.md"
+
+    @cache_response()
+    def post(self, request):
+        """Return all awards matching the provided filters and limits"""
+        models = [
+            {"name": "fields", "key": "fields", "type": "array", "array_type": "text", "text_type": "search", "min": 1},
+            {"name": "subawards", "key": "subawards", "type": "boolean", "default": False},
+        ]
+        models.extend(copy.deepcopy(AWARD_FILTER))
+        models.extend(copy.deepcopy(PAGINATION))
+        for m in models:
+            if m["name"] in ("award_type_codes", "fields"):
+                m["optional"] = False
+
+        json_request = TinyShield(models).block(request.data)
+        self.fields = json_request["fields"]
+        filters = json_request.get("filters", {})
+        subawards = json_request["subawards"]
+        self.order = json_request["order"]
+        self.limit = json_request["limit"]
+        self.page = json_request["page"]
+        self.lower_bound = (self.page - 1) * self.limit
+        self.upper_bound = self.page * self.limit + 1
+
+        if "no intersection" in filters["award_type_codes"]:
+            # "Special case": there will never be results when the website provides this value
+            return Response(
+                {"limit": self.limit, "results": [], "page_metadata": {"page": self.page, "hasNext": False}}
+            )
+
+        self.sort = json_request.get("sort") or self.fields[0]
+        if self.sort not in self.fields:
+            raise InvalidParameterException(
+                "Sort value '{}' not found in requested fields: {}".format(self.sort, self.fields)
+            )
+
+        if subawards:
+            return Response(self._handle_subawards(filters))
+
+        return Response(self._handle_awards(filters))
+
+    def _handle_awards(self, filters):
+        """ Award logic for SpendingByAward"""
+        try:
+            obtain_view_from_award_group(filters["award_type_codes"])
```

Comment:
Nothing is set to this function's return. Are you just calling it for the side-effects?

Reply:
There don't seem to be any (which is a very good thing, but confuses me more)

Reply:
The return is wrapped in a Response() for the POST above it.

Reply:
that's the return of `_handle_awards`, but I was talking about just `obtain_view_from_award_group`

Reply:
obtain_view_from_award_group() will throw an error if the type list isn't a proper subset of a category of award type codes

Reply:
That's tricky logic to follow. I would advise putting the rest of the method in a try, then catching a specific exception obtain_view_from_award_group() throws afterwards and doing error handling there. That, or building a `award_types_valid` function. I would lean slightly towards the former.

Reply:
Refactored the check